data_structures/BinaryIndexedTree/bit.py

```diff
@@ -0,0 +1,110 @@
+from __future__ import print_function
+
+__author__ = 'Nikhil Kumar'
+
+
+class BIT(object):
+    '''
+    Binary indexed tree is a simple data structure that is helpful in situations when we have lots
+    of queries which want to apply certain operation over a range of indexes in array. For example,
+    lets assume we have an array of length N. There can be two types of queries:
+        1. Change the value stored at an index i. (This is called a point update operation)
+        2. Find the sum of a prefix of length k. (This is called a range sum query)
+    In a traditional array, operation 1 would take O(1) time, while operation 2 would take O(N)
+    time in worst case. And if we have 1 million queries of type 2, we will end up wasting lot
+    of time.
+    To come out of such cases, we use binary indexed trees(BlsIT). ( segment trees is other solution )
+    I may run of space if i try to explain BIT here, but consider going to following links:
+
+    https://www.hackerearth.com/practice/data-structures/advanced-data-structures/fenwick-binary-indexed-trees/tutorial/
+    https://cs.stackexchange.com/questions/10538/bit-what-is-the-intuition-behind-a-binary-indexed-tree-and-how-was-it-thought-a
+
+    In general, it's a simple array where we store partial sums. Every index which is a power of 2,
+    stores sum of numbers from the original array from index 1 to pow(2, index). For example an index of 8 in BIT would
+    store complete sum of original array from index 1 to index 8. Then applying same logic
+    recursively to other indexes and cleverly manipulating bits in indexes, we construct and query
+    BIT. visit the links for more information.
+
+    Running time:
+        Update operation:  O(logN)
+        Query operation:  O(logN)
+
+    if you know about segment trees, you may find this data structure similar.
+    The only reason we use BIT over segment tree becasue it takes less time to code it and
+    it takes less space in memory.
+    '''
+
+    def __init__(self, arr):
+        self.input = arr
+        self.length = len(self.input)
+        self.bit = [0] * (self.length + 1)  # we will use indexes from 1 ..N
+
+    def construct(self):
```

Comment:
We probably don't need this method, it can be done in the constructor itself.

Reply:
sure will move into init.  Also i will update existing fenwick tree in BinaryTree folder. I didn't see that before. I thought Binary Indexed Tree would have it's separate folder.  it should ? shouldn't it  ?.

Reply:
Yeah, BIT can have it's own directory, we need other's opinion here. @DravitLochan @harshildarji